File: 프로그래머스/3/42898. 등굣길/등굣길.java
class Solution {
    public int solution(int m, int n, int[][] puddles) {
        int mod = 1000000007;
        int[][] dp = new int[m][n];
        for (int[] puddle : puddles) {
            dp[puddle[0] - 1][puddle[1] - 1] = -1;
        }
        
        dp[0][0] = 1;
        
        for(int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if(dp[i][j] == -1) {
                    dp[i][j] = 0;
                    continue;
                }
                
                if(i != 0) // 맨위 아님
                    dp[i][j] += dp[i - 1][j] % mod;

                if(j != 0) // 맨옆 아님
                    dp[i][j] += dp[i][j - 1] % mod;
            }
        }
        return dp[m - 1][n - 1] % mod;
    }
}

File: 프로그래머스/3/42898. 등굣길/등굣길.py
def solution(m, n, puddles):
    dp: list[list[int]] = [[0] * m for _ in range(n)]
    dp[0][0] = 1
    
    for i in range(n):
        for j in range(m):
            if [j+1, i+1] in puddles:
                continue
            if i > 0:
                dp[i][j] += dp[i-1][j]
            if j > 0:
                dp[i][j] += dp[i][j-1]
            dp[i][j] %= 1000000007
    
    return dp[n-1][m-1]